### PR TITLE
PLAT-300: pass whole state in VStateReport

### PR DESCRIPTION
**Pull request by sexyhamster**
- pass whole state in VStateReport
- pass whole objectDescriptor upon modification

### CODE REVIEW

File: ledger-core/v2/virtual/handlers/vstatereport.go [hunk 1]
```diff
@@ -45,12 +47,32 @@ func (s *SMVStateReport) Init(ctx smachine.InitializationContext) smachine.State
...
+		immutableExecuteSemaphore smsync.SemaphoreLink
+		mutableExecuteSemaphore   smsync.SemaphoreLink
+
+		deactivated *bool
+	)
+
+	if s.Payload.ProvidedContent != nil && s.Payload.ProvidedContent.LatestDirtyState != nil {
+		objectDescriptor = descriptor.NewObject(
+			objectRef,
+			s.Payload.ProvidedContent.LatestDirtyState.Reference,
```
Comment (ivanshibitov) [inline ledger-core/v2/virtual/handlers/vstatereport.go:67]: I would assign 's.Payload.ProvidedContent.LatestDirtyState' to some variable and would use this variable

File: ledger-core/v2/virtual/handlers/vstatereport.go [hunk 2]
```diff
@@ -74,5 +105,8 @@ func (s *SMVStateReport) stepProcess(ctx smachine.ExecutionContext) smachine.Sta
 		panic(throw.NotImplemented())
 	}
 
+	ctx.ApplyAdjustment(immutableExecuteSemaphore.NewDelta(immutablePendingCount))
```
Comment (ivanshibitov) [inline ledger-core/v2/virtual/handlers/vstatereport.go:108]: I thought that this should be done inside SMObject

File: ledger-core/v2/virtual/handlers/vstatereport.go [hunk 3]
```diff
@@ -45,12 +47,32 @@ func (s *SMVStateReport) Init(ctx smachine.InitializationContext) smachine.State
...
+	var (
+		objectDescriptor descriptor.Object
+
+		immutableExecuteSemaphore smsync.SemaphoreLink
+		mutableExecuteSemaphore   smsync.SemaphoreLink
+
+		deactivated *bool
+	)
+
+	if s.Payload.ProvidedContent != nil && s.Payload.ProvidedContent.LatestDirtyState != nil {
```
Comment (ivanshibitov) [inline ledger-core/v2/virtual/handlers/vstatereport.go:64]: In what case do they equal to 'nil'?

File: ledger-core/v2/virtual/execute/execute.go [hunk 4]
```diff
@@ -285,15 +287,41 @@ func (s *SMExecute) stepSaveNewObject(ctx smachine.ExecutionContext) smachine.St
 	case insolar.RequestSideEffectActivate:
 		_, prototype, memory = executionNewState.Activate()
 		action = func(state *object.SharedState) {
-			state.Info.SetDescriptor(&prototype, memory)
+			stateID := NewStateID(s.pulseSlot.PulseData().GetPulseNumber(), memory)
```
Comment (ivanshibitov) [inline ledger-core/v2/virtual/execute/execute.go:290]: It looks the same like in 'case insolar.RequestSideEffectAmend'

File: ledger-core/v2/virtual/handlers/vstatereport.go (hunk 3 above)
Comment (sexyhamster) [reply ledger-core/v2/virtual/handlers/vstatereport.go:64]: if it was not requested via VStateRequest

File: ledger-core/v2/virtual/handlers/vstatereport.go (hunk 2 above)
Comment (ruz) [reply ledger-core/v2/virtual/handlers/vstatereport.go:108]: this is SO wrong :) counters that are in VStateReport shouldn't change semaphores, they should change counters.

Comment (sexyhamster) [reply ledger-core/v2/virtual/handlers/vstatereport.go]: you are right, when I've started the ticket there were no counters so I assumed wrong behavior. Fixed

**Comment by bigbes** [comment]
👌

**Pull request merged**